be/src/olap/tablet_meta.cpp

```diff
@@ -710,4 +742,106 @@ bool operator!=(const TabletMeta& a, const TabletMeta& b) {
     return !(a == b);
 }
 
+DeleteBitmap::DeleteBitmap() {
+}
+
+DeleteBitmap::DeleteBitmap(const DeleteBitmap& o) {
+    delete_bitmap = o.delete_bitmap; // just copy data
+}
+
+DeleteBitmap& DeleteBitmap::operator=(const DeleteBitmap& o) {
+    delete_bitmap = o.delete_bitmap; // just copy data
+    return *this;
+}
+
+DeleteBitmap::DeleteBitmap(DeleteBitmap&& o) {
+    delete_bitmap = std::move(o.delete_bitmap);
+}
+
+DeleteBitmap& DeleteBitmap::operator=(DeleteBitmap&& o) {
+    delete_bitmap = std::move(o.delete_bitmap);
+    return *this;
+}
+
+DeleteBitmap DeleteBitmap::snapshot() const {
+    std::shared_lock l(lock);
+    return DeleteBitmap(*this);
+}
+
+void DeleteBitmap::add(const BitmapKey& bitmap, uint32_t row_id) {
+    std::lock_guard l(lock);
```

Comment:
It looks like the granularity of the lock is tablet level.

I wonder will there be a performance penalty for this part of serial update when loading concurrently that need to update multiple versions of bitmaps?

Reply:
Your concern makes sense, however, 
1. This `add` operation is in-memory and fairly fast enough, and contention on this lock is not significant because there won't be thousands of concurrent loads on a single tablet.
2. The read lock is acquired when the tablet meta is being saved (called by `TabletManager::do_tablet_meta_checkpoint`), serialized bitmap will go to RocksDB, more precisely -- page cache of WAL, it's also kind of memory operation.

It is OK that we don't optimize the lock granularity for now.